lib/docker/connection.rb

```diff
@@ -35,21 +35,51 @@ def resource
 
   # Send a request to the server with the `
   def request(*args, &block)
+    retries ||= 0
     request = compile_request_params(*args, &block)
     log_request(request)
-    resource.request(request).body
-  rescue Excon::Errors::BadRequest => ex
-    raise ClientError, ex.response.body
-  rescue Excon::Errors::Unauthorized => ex
-    raise UnauthorizedError, ex.response.body
-  rescue Excon::Errors::NotFound => ex
-    raise NotFoundError, ex.response.body
-  rescue Excon::Errors::Conflict => ex
-    raise ConflictError, ex.response.body
-  rescue Excon::Errors::InternalServerError => ex
-    raise ServerError, ex.response.body
-  rescue Excon::Errors::Timeout => ex
-    raise TimeoutError, ex.message
+    begin
+      resource.request(request).body
+    rescue Excon::Errors::BadRequest => ex
+      if retries < 2
+        response_cause = ''
+        begin
+          response_cause = JSON.parse(ex.response.body)['cause']
+        rescue JSON::ParserError
+          #noop
+        end
+
+        if response_cause.is_a?(String)
+          # The error message will tell the application type given and then the
+          # application type that the message should be
+          #
+          # This is not perfect since it relies on processing a message that
+          # could change in the future. However, it should be a good stop-gap
+          # until all methods are updated to pass in the appropriate content
+          # type.
+          #
+          # A current example message is:
+          #   * 'Content-Type: application/json is not supported. Should be "application/x-tar"'
+          matches = response_cause.delete('"\'').scan(%r{(application/\S+)})
+          unless matches.count < 2
+            request[:headers]['Content-Type'] = matches.last.first
```

Comment:
This is pretty spooky to me. Can we log something so that folks can clearly open an issue or something when this is hit? Also, is there a particular endpoint that you know needs to be fixed?

Reply:
Not going to lie, it's 100% spooky. I'm fine with logging things if you have a preferred logging mechanism. I allowed updates from maintainers so feel free to hack away if that's easier and I can put it back through its paces in podman.

I took a swipe at the Docker API docs but multiple endpoints need to be updated and the downstream code doesn't actually support passing it in as an option.

As the comment mentions, this is basically a stop-gap until the real fix can be introduced later (which will probably involve a larger code overhaul if I'm reading things right).

Reply:
I guess the limited ability for logging is from: https://github.com/swipely/docker-api/pull/569/files#diff-a8fbee6483c191f77ba681f51094f68035815545d435d51a195934263bd79850R86-R90

Doesn't really offer much since it's opt in...

Reply:
Also, while this is spooky, it's "safe spooky" in that the worst case scenario is that the thing you were trying to do gets tried twice and doesn't work (and returns the usual error).

Reply:
My fear mostly stems from the docker engine API being a bit harder to work with earlier on and potentially performing destructive behaviors, but not returning the right result. Agreed that it shouldn't actually do that here since it's failing at the API top-level.

Reply:
Well, `docker` doesn't care. It seems to just ignore it and "do the right thing" for the operation. The `podman` API is strict and will fail.